File: Assignment-3/CPP/AEState.cpp
//===- AEState.cpp -- Assignment-3 abstract-state facade ------------------//

#include "AEState.h"
#include "SVFIR/SVFIR.h"
#include "Util/Options.h"

#include <cassert>
#include <cstdlib>

namespace SVF {

	AbstractValue& AEState::load(u32_t addr) {
		return state.load(addr);
	}

	const AbstractValue& AEState::load(u32_t addr) const {
		return state.load(addr);
	}

	void AEState::store(u32_t addr, const AbstractValue& value) {
		state.store(addr, value);
	}

	const AbstractValue& AEState::getAbsValue(const ValVar* var) const {
		u32_t id = var->getId();
		AbstractState& mutableState = const_cast<AbstractState&>(state);
		if (!mutableState.getVarToVal().count(id))
			mutableState[id] = IntervalValue::top();
		return mutableState[id];
	}

	const AbstractValue& AEState::getAbsValue(const ObjVar* var) const {
		AbstractState& mutableState = const_cast<AbstractState&>(state);
		return mutableState.load(AbstractState::getVirtualMemAddress(var->getId()));
	}

	const AbstractValue& AEState::getAbsValue(const SVFVar* var) const {
		if (const ObjVar* obj = SVFUtil::dyn_cast<ObjVar>(var))
			return getAbsValue(obj);
		if (const ValVar* value = SVFUtil::dyn_cast<ValVar>(var))
			return getAbsValue(value);
		assert(false && "Unknown SVFVar kind");
		std::abort();
	}

	void AEState::updateAbsValue(const ValVar* var, const AbstractValue& value) {
		state[var->getId()] = value;
	}

	void AEState::updateAbsValue(const ObjVar* var, const AbstractValue& value) {
		state.store(AbstractState::getVirtualMemAddress(var->getId()), value);
	}

	void AEState::updateAbsValue(const SVFVar* var, const AbstractValue& value) {
		if (const ObjVar* obj = SVFUtil::dyn_cast<ObjVar>(var))
			updateAbsValue(obj, value);
		else if (const ValVar* val = SVFUtil::dyn_cast<ValVar>(var))
			updateAbsValue(val, value);
		else
			assert(false && "Unknown SVFVar kind");
	}

	const ValVar* AEState::getValVar(NodeID id) const {
		const SVFVar* var = PAG::getPAG()->getGNode(id);
		assert(SVFUtil::isa<ValVar>(var) && "AEState operation requires a ValVar");
		return SVFUtil::cast<ValVar>(var);
	}

	AbstractValue AEState::loadValue(const ValVar* pointer) const {
		AbstractValue result;
		const AbstractValue& ptr = getAbsValue(pointer);
		for (u32_t addr : ptr.getAddrs()) {
			const SVFVar* object = PAG::getPAG()->getSVFVar(state.getIDFromAddr(addr));
			result.join_with(getAbsValue(object));
		}
		return result;
	}

	AbstractValue AEState::loadValue(NodeID pointerId) const {
		return loadValue(getValVar(pointerId));
	}

	void AEState::storeValue(const ValVar* pointer, const AbstractValue& value) {
		const AbstractValue& ptr = getAbsValue(pointer);
		for (u32_t addr : ptr.getAddrs()) {
			const SVFVar* object = PAG::getPAG()->getSVFVar(state.getIDFromAddr(addr));
			updateAbsValue(object, value);
		}
	}

	void AEState::storeValue(NodeID pointerId, const AbstractValue& value) {
		storeValue(getValVar(pointerId), value);
	}

	AddressValue AEState::getGepObjAddrs(const ValVar* pointer, IntervalValue offset) {
		AddressValue addresses;
		APOffset lb = offset.lb().getIntNumeral() < Options::MaxFieldLimit()
		              ? offset.lb().getIntNumeral() : Options::MaxFieldLimit();
		APOffset ub = offset.ub().getIntNumeral() < Options::MaxFieldLimit()
		              ? offset.ub().getIntNumeral() : Options::MaxFieldLimit();
		for (APOffset index = lb; index <= ub; ++index) {
			for (u32_t address : getAbsValue(pointer).getAddrs()) {
				NodeID base = state.getIDFromAddr(address);
				assert(SVFUtil::isa<ObjVar>(PAG::getPAG()->getSVFVar(base)) &&
				       "Fail to get the base object address");
				NodeID gepObj = PAG::getPAG()->getGepObjVar(base, index);
				state[gepObj] = AddressValue(AbstractState::getVirtualMemAddress(gepObj));
				addresses.insert(AbstractState::getVirtualMemAddress(gepObj));
			}
		}
		return addresses;
	}

	AddressValue AEState::getGepObjAddrs(NodeID pointerId, IntervalValue offset) {
		return getGepObjAddrs(getValVar(pointerId), offset);
	}

	const SVFType* AEState::getPointeeElement(const ObjVar* object) const {
		if (!object) return nullptr;
		const AbstractValue& ptr = getAbsValue(object);
		if (!ptr.isAddr()) return nullptr;
		for (u32_t address : ptr.getAddrs()) {
			NodeID objectId = state.getIDFromAddr(address);
			if (objectId != 0)
				return PAG::getPAG()->getBaseObject(objectId)->getType();
		}
		return nullptr;
	}

	IntervalValue AEState::getGepElementIndex(const GepStmt* gep) const {
		if (gep->isConstantOffset())
			return IntervalValue((s64_t)gep->accumulateConstantOffset());
		IntervalValue result(0);
		for (int i = (int)gep->getOffsetVarAndGepTypePairVec().size() - 1; i >= 0; --i) {
			const ValVar* var = gep->getOffsetVarAndGepTypePairVec()[i].first;
			const SVFType* type = gep->getOffsetVarAndGepTypePairVec()[i].second;
			s64_t lb, ub;
			if (const ConstIntValVar* constant = SVFUtil::dyn_cast<ConstIntValVar>(var))
				lb = ub = constant->getSExtValue();
			else {
				IntervalValue interval = getAbsValue(var).getInterval();
				if (interval.isBottom()) lb = ub = 0;
				else { lb = interval.lb().getIntNumeral(); ub = interval.ub().getIntNumeral(); }
			}
			if (SVFUtil::isa<SVFPointerType>(type)) {
				u32_t count = gep->getAccessPath().getElementNum(gep->getAccessPath().gepSrcPointeeType());
				lb = (double)Options::MaxFieldLimit() / count < lb ? Options::MaxFieldLimit() : lb * count;
				ub = (double)Options::MaxFieldLimit() / count < ub ? Options::MaxFieldLimit() : ub * count;
			} else if (Options::ModelArrays()) {
				const std::vector<u32_t>& flat = PAG::getPAG()->getTypeInfo(type)->getFlattenedElemIdxVec();
				if (flat.empty() || ub >= (APOffset)flat.size() || lb < 0) lb = ub = 0;
				else { lb = PAG::getPAG()->getFlattenedElemIdx(type, lb); ub = PAG::getPAG()->getFlattenedElemIdx(type, ub); }
			} else lb = ub = 0;
			result = result + IntervalValue(lb, ub);
		}
		result.meet_with(IntervalValue((s64_t)0, (s64_t)Options::MaxFieldLimit()));
		return result.isBottom() ? IntervalValue(0) : result;
	}

	IntervalValue AEState::getGepByteOffset(const GepStmt* gep) const {
		if (gep->isConstantOffset())
			return IntervalValue((s64_t)gep->accumulateConstantByteOffset());
		IntervalValue result(0);
		for (int i = (int)gep->getOffsetVarAndGepTypePairVec().size() - 1; i >= 0; --i) {
			const ValVar* var = gep->getOffsetVarAndGepTypePairVec()[i].first;
			const SVFType* type = gep->getOffsetVarAndGepTypePairVec()[i].second;
			if (SVFUtil::isa<SVFArrayType>(type) || SVFUtil::isa<SVFPointerType>(type)) {
				u32_t size = SVFUtil::isa<SVFArrayType>(type)
				             ? SVFUtil::cast<SVFArrayType>(type)->getTypeOfElement()->getByteSize()
				             : gep->getAccessPath().gepSrcPointeeType()->getByteSize();
				if (const ConstIntValVar* constant = SVFUtil::dyn_cast<ConstIntValVar>(var)) {
					s64_t value = (double)Options::MaxFieldLimit() / size >= constant->getSExtValue()
					              ? constant->getSExtValue() * size : Options::MaxFieldLimit();
					result = result + IntervalValue(value, value);
				} else {
					IntervalValue index = getAbsValue(var).getInterval();
					if (index.isBottom()) result = result + IntervalValue(0, 0);
					else {
						auto scale = [size](s64_t value) {
							if (value < 0) return (s64_t)0;
							return (double)Options::MaxFieldLimit() / size >= value
							       ? (s64_t)size * value : (s64_t)Options::MaxFieldLimit();
						};
						result = result + IntervalValue(scale(index.lb().getIntNumeral()), scale(index.ub().getIntNumeral()));
					}
				}
			} else if (const SVFStructType* structure = SVFUtil::dyn_cast<SVFStructType>(type)) {
				result = result + IntervalValue(gep->getAccessPath().getStructFieldOffset(var, structure));
			} else assert(false && "gep type pair only supports array/pointer/struct");
		}
		return result;
	}

	u32_t AEState::getAllocaInstByteSize(const AddrStmt* addr) const {
		const ObjVar* object = SVFUtil::dyn_cast<ObjVar>(addr->getRHSVar());
		assert(object && "Addr rhs value is not ObjVar");
		const BaseObjVar* base = PAG::getPAG()->getBaseObject(object->getId());
		if (base->isConstantByteSize()) return base->getByteSizeOfObj();
		u64_t result = 1;
		for (const SVFVar* size : addr->getArrSize()) {
			IntervalValue interval = getAbsValue(size).getInterval();
			if (interval.isBottom()) interval = IntervalValue(Options::MaxFieldLimit());
			result = result * interval.ub().getIntNumeral() > Options::MaxFieldLimit()
			         ? Options::MaxFieldLimit() : result * interval.ub().getIntNumeral();
		}
		return (u32_t)result;
	}

	void AEState::joinWith(const AbstractState& other) {
		state.joinWith(other);
	}

	AbstractState AEState::widening(const AbstractState& other) const {
		return const_cast<AbstractState&>(raw()).widening(other);
	}

	AbstractState AEState::narrowing(const AbstractState& other) const {
		return const_cast<AbstractState&>(raw()).narrowing(other);
	}

	bool AEState::equals(const AbstractState& other) const {
		return raw().equals(other);
	}

	void AEState::printAbstractState() const {
		raw().printAbstractState();
	}

} // namespace SVF

File: Assignment-3/Python/AEState.py
"""Assignment-3 abstract-state facade."""

from typing import Any

import pysvf


def unwrap_state(state):
    """Return the PySVF state behind an Assignment-3 AEState."""
    return state.raw if isinstance(state, AEState) else state


class AEState:
    """Assignment-3-owned wrapper around one PySVF AbstractState."""

    isBlackHoleObjAddr = staticmethod(pysvf.AbstractState.isBlackHoleObjAddr)
    isNullMem = staticmethod(pysvf.AbstractState.isNullMem)
    isVirtualMemAddress = staticmethod(pysvf.AbstractState.isVirtualMemAddress)
    getVirtualMemAddress = staticmethod(pysvf.AbstractState.getVirtualMemAddress)

    def __init__(self, state=None) -> None:
        self._state = state if state is not None else pysvf.AbstractState()

    @property
    def raw(self) -> pysvf.AbstractState:
        return self._state

    def __getitem__(self, var_id: int) -> pysvf.AbstractValue:
        return self.raw[var_id]

    def __setitem__(self, var_id: int, value) -> None:
        self.raw[var_id] = value

    def __getattr__(self, name: str) -> Any:
        return getattr(self.raw, name)

    def __eq__(self, other) -> bool:
        return self.raw == unwrap_state(other)

    def __str__(self) -> str:
        return str(self.raw)

    def __repr__(self) -> str:
        return f"AEState(state={self.raw!r})"

    @staticmethod
    def _resolve_var(var):
        return pysvf.getPAG().getGNode(var) if isinstance(var, int) else var

    def getAbsValue(self, var: pysvf.SVFVar) -> pysvf.AbstractValue:
        if isinstance(var, pysvf.ObjVar):
            addr = pysvf.AbstractState.getVirtualMemAddress(var.getId())
            return self.raw.load(addr)
        values = self.raw.getVarToVal()
        if var.getId() not in values:
            self.raw[var.getId()] = pysvf.AbstractValue(pysvf.IntervalValue.top())
        return self.raw[var.getId()]

    def updateAbsValue(self, var: pysvf.SVFVar, value) -> None:
        if isinstance(var, pysvf.ObjVar):
            addr = pysvf.AbstractState.getVirtualMemAddress(var.getId())
            self.raw.store(addr, value)
        else:
            self.raw[var.getId()] = value

    def loadValue(self, pointer) -> pysvf.AbstractValue:
        pointer = self._resolve_var(pointer)
        result = pysvf.AbstractValue()
        value = self.getAbsValue(pointer)
        for address in value.getAddrs():
            obj = pysvf.getPAG().getGNode(self.raw.getIDFromAddr(address))
            result.join_with(self.getAbsValue(obj))
        return result

    def storeValue(self, pointer, value: pysvf.AbstractValue) -> None:
        pointer = self._resolve_var(pointer)
        value_at_pointer = self.getAbsValue(pointer)
        for address in value_at_pointer.getAddrs():
            obj = pysvf.getPAG().getGNode(self.raw.getIDFromAddr(address))
            self.updateAbsValue(obj, value)

    def getGepObjAddrs(
        self,
        pointer,
        offset: pysvf.IntervalValue,
    ) -> pysvf.AddressValue:
        pointer = self._resolve_var(pointer)
        pag = pysvf.getPAG()
        limit = pysvf.Options.max_field_limit()
        lower = min(offset.lb().getNumeral(), limit)
        upper = min(offset.ub().getNumeral(), limit)
        addresses = pysvf.AddressValue(set())
        for index in range(lower, upper + 1):
            for address in self.getAbsValue(pointer).getAddrs():
                base_id = self.raw.getIDFromAddr(address)
                base = pag.getGNode(base_id)
                assert isinstance(base, pysvf.ObjVar), "Fail to get the base object address"
                gep_obj = pag.getGepObjVar(base_id, index)
                gep_addr = pysvf.AbstractState.getVirtualMemAddress(gep_obj)
                self.raw[gep_obj] = pysvf.AbstractValue(pysvf.AddressValue(gep_addr))
                addresses.insert(gep_addr)
        return addresses

    def getGepElementIndex(self, gep: pysvf.GepStmt) -> pysvf.IntervalValue:
        if gep.isConstantOffset():
            return pysvf.IntervalValue(gep.getConstantOffset())
        result = pysvf.IntervalValue(0)
        limit = pysvf.Options.max_field_limit()
        pag = pysvf.getPAG()
        for var, value_type in reversed(gep.getOffsetVarAndGepTypePairVec()):
            if var.isConstIntValVar():
                lower = upper = var.asConstIntValVar().getSExtValue()
            else:
                interval = self.getAbsValue(var).getInterval()
                if interval.isBottom():
                    lower = upper = 0
                else:
                    lower = interval.lb().getNumeral()
                    upper = interval.ub().getNumeral()
            if value_type.isPointerType():
                count = pag.getNumOfFlattenElements(gep.getSrcPointeeType())
                lower = max(0, min(limit, lower * count))
                upper = max(0, min(limit, upper * count))
            else:
                count = pag.getNumOfFlattenElements(value_type)
                if count == 0 or upper >= count or lower < 0:
                    lower = upper = 0
                else:
                    lower = pag.getFlattenedElemIdx(value_type, lower)
                    upper = pag.getFlattenedElemIdx(value_type, upper)
            result = result + pysvf.IntervalValue(lower, upper)
        result.meet_with(pysvf.IntervalValue(0, limit))
        return pysvf.IntervalValue(0) if result.isBottom() else result

    def getElementIndex(self, gep: pysvf.GepStmt) -> pysvf.IntervalValue:
        return self.getGepElementIndex(gep)

    def getGepByteOffset(self, gep: pysvf.GepStmt) -> pysvf.IntervalValue:
        if gep.isConstantOffset():
            return pysvf.IntervalValue(gep.getConstantByteOffset())
        result = pysvf.IntervalValue(0)
        limit = pysvf.Options.max_field_limit()
        for var, value_type in reversed(gep.getOffsetVarAndGepTypePairVec()):
            if value_type.isArrayType() or value_type.isPointerType():
                size = (value_type.asArrayType().getTypeOfElement().getByteSize()
                        if value_type.isArrayType()
                        else gep.getSrcPointeeType().getByteSize())
                if var.isConstIntValVar():
                    value = var.asConstIntValVar().getSExtValue()
                    scaled = value * size if limit / size >= value else limit
                    result = result + pysvf.IntervalValue(scaled)
                else:
                    index = self.getAbsValue(var).getInterval()
                    if index.isBottom():
                        result = result + pysvf.IntervalValue(0)
                    else:
                        def scale(value):
                            if value < 0:
                                return 0
                            return value * size if limit / size >= value else limit
                        result = result + pysvf.IntervalValue(
                            scale(index.lb().getNumeral()),
                            scale(index.ub().getNumeral()))
            elif value_type.isStructType():
                result = result + pysvf.IntervalValue(
                    gep.getStructFieldOffset(var, value_type.asStructType()))
            else:
                raise AssertionError("GEP type must be array, pointer, or struct")
        return result

    def getByteOffset(self, gep: pysvf.GepStmt) -> pysvf.IntervalValue:
        return self.getGepByteOffset(gep)

    def getAllocaInstByteSize(self, addr: pysvf.AddrStmt) -> int:
        obj = addr.getRHSVar()
        assert isinstance(obj, pysvf.ObjVar), "Addr rhs value is not ObjVar"
        base = pysvf.getPAG().getBaseObject(obj.getId())
        if base.isConstantByteSize():
            return base.getByteSizeOfObj()
        result = 1
        limit = pysvf.Options.max_field_limit()
        for size_var in addr.getArrSize():
            interval = self.getAbsValue(size_var).getInterval()
            if interval.isBottom():
                interval = pysvf.IntervalValue(limit)
            size = interval.ub().getNumeral()
            result = limit if result * size > limit else result * size
        return result

    def getPointeeElement(self, pointer_id: int):
        for address in self.raw[pointer_id].getAddrs():
            obj_id = self.raw.getIDFromAddr(address)
            pag = pysvf.getPAG()
            obj = pag.getGNode(obj_id)
            base_object = pag.getBaseObject(obj_id)
            if base_object is not None:
                return base_object.getType()
        return None

    def joinWith(self, other) -> None:
        self.raw.joinWith(unwrap_state(other))

    def widening(self, other) -> pysvf.AbstractState:
        return self.raw.widening(unwrap_state(other))

    def narrowing(self, other) -> pysvf.AbstractState:
        return self.raw.narrowing(unwrap_state(other))

    def equals(self, other) -> bool:
        return self.raw.equals(unwrap_state(other))
